src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
+                      : chroma_final;
+
+  Ych_final[1] = fmaxf(chroma_final / Ych_final[0], 0.f);
+}
+
+// Pipeline and ICC luminance is CIE Y 1931
+// Kirk Ych/Yrg uses CIE Y 2006
+// 1 CIE Y 1931 = 0.945302269001 CIE Y 2006, so we need to adjust that
+#define CIE_Y_1931_to_CIE_Y_2006(x) 0.945302269001f * x
+
+
+static inline void gamut_check_RGB(dt_aligned_pixel_t RGB_in, dt_aligned_pixel_t Ych_in,
+                                   const dt_colormatrix_t matrix_in, const dt_colormatrix_t matrix_out,
+                                   const float display_black, const float display_white)
+{
+  const int max_iter = 8;
+
+  int iter = 0;
+  float max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  while(max_pix > display_white && iter < max_iter)
+  {
+    // Perform exposure correction to fit back in gamut
+    // This gives us the closest in-gamut pixel at same saturation.
+    dt_aligned_pixel_t RGB_darkened = { 0.f };
+    for_each_channel(c, aligned(RGB_darkened)) RGB_darkened[c] = RGB_in[c] / max_pix;
+    dt_aligned_pixel_t Ych_darkened = { 0.f };
+    pipe_RGB_to_Ych(RGB_darkened, matrix_in, Ych_darkened);
+
+    // Remapping goal:
+    // 1. at constant hue and luminance, march toward the chroma of the closest same-saturation in-gamut color.
+    //    Since same-saturation is darker, its chroma should be much lower than the current.
+    // 2. iterate until convergence.
+    Ych_in[1] = Ych_darkened[1] * Ych_darkened[0] / Ych_in[0];
+    Ych_to_pipe_RGB(Ych_in, matrix_out, RGB_in);
+    max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+    iter++;
+  }
+
+  // Finally, look for negatives.
+  float min_pix = fminf(fminf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  iter = 0;
+  while(min_pix < display_black && iter < max_iter)
+  {
+    // Perform a black offset to fit back in gamut
+    // This gives us a whitest version of the current color.
+    // It's not saturation-invariant.
+    dt_aligned_pixel_t RGB_brightened = { 0.f };
+    for_each_channel(c, aligned(RGB_brightened))
+      RGB_brightened[c] = RGB_in[c] + fabsf(min_pix);
+    dt_aligned_pixel_t Ych_brightened = { 0.f };
+    pipe_RGB_to_Ych(RGB_brightened, matrix_in, Ych_brightened);
+
+    // Get the locus of the clamped pixel.
+    // NB: that's where we land if we do nothing.
+    // This gives us the closest in-gamut pixel with no color property preserved.
+    dt_aligned_pixel_t RGB_clamped = { 0.f };
+    for_each_channel(c, aligned(RGB_clamped))
+      RGB_clamped[c] = CLAMP(RGB_in[c], display_black, display_white);
+    dt_aligned_pixel_t Ych_clamped = { 0.f };
+    pipe_RGB_to_Ych(RGB_clamped, matrix_in, Ych_clamped);
+
+    // Same logic as above, except we need to scale down the chroma
+    // from the brightnened projection
+    if(Ych_brightened[0] > CIE_Y_1931_to_CIE_Y_2006(display_black))
+    {
+      Ych_in[0] = fmax((Ych_in[0] + Ych_clamped[0]) / 2.f, CIE_Y_1931_to_CIE_Y_2006(display_black));
```

Comment:
Use `fmaxf`. However for better vectorizability, the `MIN()` and `MAX()` macros should be favoured. I think it's the `-fno-finite-math-only` flag in the build that causes these `fmaxf` and `fminf` to not auto-vectorize properly. Not a big deal though.

Reply:
right